IdentityCore/src/validation/MSIDAdfsAuthorityResolver.m

```diff
@@ -0,0 +1,182 @@
+        return;
+    }
+    
+    // Check for upn suffix
+    NSString *domain = [self getDomain:upn];
+    if ([NSString msidIsStringNilOrBlank:domain])
+    {
+        __auto_type error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"'upn' is a required parameter and must not be nil or empty.", nil, nil, nil, context.correlationId, nil);
+        
+        if (completionBlock) completionBlock(nil, NO, error);
+        return;
+    }
+    
+    [self sendDrsDiscoveryWithDomain:domain context:context completionBlock:^(NSURL *issuer, NSError *error)
+     {
+         __auto_type webFingerRequest = [[MSIDWebFingerRequest alloc] initWithIssuer:issuer
+                                                                           authority:authority];
+         [webFingerRequest sendWithBlock:^(id response, NSError *error)
+          {
+              if (error)
+              {
+                  if (completionBlock) completionBlock(nil, NO, error);
+                  return;
+              }
+              
+              if ([self isRealmTrustedFromWebFingerPayload:response authority:authority])
+              {
+                  __auto_type openIdConfigurationEndpoint = [self openIdConfigurationEndpointForAuthority:authority];
+                  
+                  __auto_type cacheRecord = [MSIDAuthorityCacheRecord new];
+                  cacheRecord.validated = YES;
+                  cacheRecord.openIdConfigurationEndpoint = openIdConfigurationEndpoint;
+                  [s_cache setObject:cacheRecord forKey:authority.absoluteString.lowercaseString];
+                  
+                  if (completionBlock) completionBlock(openIdConfigurationEndpoint, YES, nil);
+              }
+              else
+              {
+                  error = MSIDCreateError(MSIDErrorDomain, MSIDErrorDeveloperAuthorityValidation, @"WebFinger request was invalid or failed", nil, nil, nil, context.correlationId, nil);
+                  if (completionBlock) completionBlock(nil, NO, error);
+              }
+          }];
+     }];
+}
+
+#pragma mark - Private
+
+- (void)sendDrsDiscoveryWithDomain:(NSString *)domain
+                           context:(id<MSIDRequestContext>)context
+                   completionBlock:(MSIDHttpRequestDidCompleteBlock)completionBlock
+{
+    __auto_type drsPremRequest = [[MSIDDRSDiscoveryRequest alloc] initWithDomain:domain adfsType:MSIDADFSTypeOnPrems];
+    drsPremRequest.context = context;
+    [drsPremRequest sendWithBlock:^(id response, NSError *error)
+     {
+         if (response)
+         {
+             if (completionBlock) completionBlock(response, error);
+             return;
+         }
+         
+         __auto_type drsCloudRequest = [[MSIDDRSDiscoveryRequest alloc] initWithDomain:domain adfsType:MSIDADFSTypeCloud];
+         drsCloudRequest.context = context;
+         [drsCloudRequest sendWithBlock:^(id response, NSError *error)
+          {
+              if (response)
```

Comment:
what if there's no response for the cloud request? How will the completionBlock ever get called?

Reply:
fixed